### ISSUE TITLE
Saving ItemSimilarityRecommender to a Cloud Object Storage

### ISSUE THREAD

**Issue by dogahe** [apple/turicreate#3283]
I am trying to save a ItemSimilarityRecommender model to a Cloud Object Storage. But I am not sure, how I need to send the model to this API. Currently I have this code:

```
model_loaded = tc.load_model("MyModel")
headers = { 'Authorization': 'bearer ' + access_token, 
          'Content-Type': 'multipart/form-data' }. # I don't think I have chosen Content-Type correctly

url = 'https://some_IBM_Cloud_Object_Storage_URL/model01.data'
response = requests.put(url, data = model_loaded, headers = headers)
```

But I get this error:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-13-c7007ff209f1> in <module>
----> 1 response = requests.put(url, data = model_loaded, headers = headers)

~/anaconda3/envs/venv/lib/python3.7/site-packages/requests/api.py in put(url, data, **kwargs)
    129     """
    130 
--> 131     return request('put', url, data=data, **kwargs)
    132 
    133 

~/anaconda3/envs/venv/lib/python3.7/site-packages/requests/api.py in request(method, url, **kwargs)
     58     # cases, and look like a memory leak in others.
     59     with sessions.Session() as session:
---> 60         return session.request(method=method, url=url, **kwargs)
     61 
     62 

~/anaconda3/envs/venv/lib/python3.7/site-packages/requests/sessions.py in request(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)
    531         }
    532         send_kwargs.update(settings)
--> 533         resp = self.send(prep, **send_kwargs)
    534 
    535         return resp

~/anaconda3/envs/venv/lib/python3.7/site-packages/requests/sessions.py in send(self, request, **kwargs)
    644 
    645         # Send the request
--> 646         r = adapter.send(request, **kwargs)
    647 
    648         # Total elapsed time of the request (approximately)

~/anaconda3/envs/venv/lib/python3.7/site-packages/requests/adapters.py in send(self, request, stream, timeout, verify, cert, proxies)
    467                     low_conn.endheaders()
    468 
--> 469                     for i in request.body:
    470                         low_conn.send(hex(len(i))[2:].encode('utf-8'))
    471                         low_conn.send(b'\r\n')

TypeError: 'ItemSimilarityRecommender' object is not iterable
```

Or if I choose to use the Python SDK provided https://cloud.ibm.com/docs/cloud-object-storage/libraries?topic=cloud-object-storage-python#python-examples-multipart

And use the function provided in the above link:

multi_part_upload("my-bucket", "my_model", "MyModel")

I get the error:
Starting file transfer for my_model to bucket: my-bucket

Unable to complete multi-part upload: [Errno 21] Is a directory: 'MyModel'

**Since the model is a directory not a file.**

**Comment by TobyRoseman** [comment]
TuriCreate models are disk backed. So even if you can get all of the `model_loaded` variable copied to the cloud, it's still not going to work.

We do have built in S3 support, but it looks like you're using a different cloud provider.

If the model will only be used for inference and will run on an Apple device, you could export to CoreML and uploading that. 

I think your best option is just copying the saved model files to the cloud.

**Comment by dogahe** [comment]
Hi - I think IBM Cloud Object Storage is compatible with S3. So if you tell me how you do it in S3 I may be able to figure out how to do it on IBM Cloud Object Storage? I would like to have the model in our backend since we can use it with other platforms like web as well. When you said "I think your best option is just copying the saved model files to the cloud.", can you please elaborate more? You mean I upload each file (.ini, .sidx, .bin , .0000) individually? And when I need to load the model download them individually, put them in a folder and call tc.load_model on it?

If you could elaborate on the two items, I'd appreciate that. 1. S3 2. Copying saved model files.

**Comment by TobyRoseman** [comment]
To use S3 - first set your AWS credentials in your environment. I usually execute the following system command before starting my Python session:
```
export AWS_ACCESS_KEY_ID=XXXX
export AWS_SECRET_ACCESS_KEY=XXXX
```
Of course replace `XXXX` with your values.

Once you've started Python and imported turicreate. You'll probably want to set `TURI_S3_ENDPOINT` and `TURI_S3_REGION` via [turicreate.config.set_runtime_config](https://apple.github.io/turicreate/docs/api/generated/turicreate.config.set_runtime_config.html#turicreate.config.set_runtime_config).

Then all you should need to do is call `my_model.save("s3:\\<bucket_name>\<key>")`.

----------

It sounds like you already understand the copying approach I mentioned; yes, you'll want to copy the (.ini, .sidx, .bin , .0000) individual files. I would probably tar the entire directory and then just copy the tar file.

**Comment by TobyRoseman** [comment]
@dogahe - I'm going to close this issue but let me know if you have any additional questions.

**Comment by dogahe** [comment]
@TobyRoseman Hi - I gave this a shot. I was able to find out some of the settings in IBM Cloud Object Storage that allows me to set the parameters as you suggested. However, the last line of code you suggested (my_model.save("s3:\\my-bucket\my-model")) just saved a local model file at folder s3/     did I understand something incorrectly?

**Comment by TobyRoseman** [comment]
Do thing work outside of the notebook?

Our S3 code doesn't have any progress reporting. So it's possible it's just taking taking a while and not actually hanging. Try saving and load a very small model:
```python
import turicreate as tc

S3_PATH = 's3://bucket-name/model-name'

sf = tc.SFrame({'x': [1,2], 'y': [2,4]})
model = tc.linear_regression.create(sf, 'y')
model.save(S3_PATH)
model = tc.load_model(S3_PATH)
```

**Comment by dogahe** [comment]
@TobyRoseman 
Thanks, for the small model you suggested it worked. I yet have to find out how long is the wait for my real model which could be between 50-100MB. Past few time I ran to load the model from the IBM Cloud Object Storage I gave up after waiting a long time 20-30 minutes. I don't know where the delay comes from. Whether some on the IBM Cloud Object Storage or TuriCreate, not sure how I can find out. Because I can't imagine it would take more than what I already waited for for a 50MB model.
